Create certificates, admin kubeconfig, bootstrap token and discoveryhashes using kubeadm

awesomenix opened the pull request:
**What this PR does / why we need it**:

Currently we manually generate certificates and supporting configuration, instead use kubeadm standard tooling for generating all the base configuration required to bootstrap kubernetes nodes

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Usage of upstream kubeadm tool for generation of prerequisites
```

awesomenix commented:
/assign @tariq1890 @justaugustus

tariq1890 commented on Gopkg.toml:
```diff
@@ -53,9 +53,51 @@ required = [
+  name="sigs.k8s.io/controller-tools"
+  version="v0.1.1"
+
+# For dependency below: Refer to issue https://github.com/golang/dep/issues/1799
```
This comment actually refers to the `fsnotify` if I am not wrong

tariq1890 commented on Gopkg.toml:
```diff
@@ -53,9 +53,51 @@ required = [
+  version="v0.1.1"
+
+# For dependency below: Refer to issue https://github.com/golang/dep/issues/1799
+[[override]]
```
I would highly advise against `override` deps. It will cause several problems. I had discussed this with @justaugustus as well.

tariq1890 replied:
I have this PR so that we can import the latest k8s the right way : https://github.com/kubernetes-sigs/controller-tools/pull/164

awesomenix replied:
Thanks Tariq, removed all overrides

tariq1890 commented on Gopkg.toml:
```diff
@@ -53,9 +53,51 @@ required = [
+# For dependency below: Refer to issue https://github.com/golang/dep/issues/1799
+[[override]]
+  name = "k8s.io/kubernetes"
+  version = "=v1.13.3"
```
We should be using `1.13.4` if we were to version pin at all.

awesomenix replied:
Yep! removed all overrides

tariq1890 commented on pkg/cloud/azure/services/compute/virtualmachines.go:
```diff
@@ -167,7 +161,7 @@ func (s *Service) getVMStartupScript(machine *actuators.MachineScope, bootstrapT
 				CACert:            string(s.scope.ClusterConfig.CAKeyPair.Cert),
 				CAKey:             string(s.scope.ClusterConfig.CAKeyPair.Key),
-				CACertHash:        caCertHash,
+				CACertHash:        s.scope.ClusterConfig.DiscoveryHashes[0],
```
We should do a len check before referencing an index explicitly.

justaugustus commented:
/hold
(For my review)

I'm at a conference this week, so my availability is going to be limited.

justaugustus commented:
/ok-to-test

awesomenix commented:
Thanks @tariq1890 i have removed all the overrides, great suggestion (its much simpler now). Fixed all the lint and test issues, could you please take another look.

tariq1890 commented on pkg/cloud/azure/services/certificates/certificates_test.go:
```diff
@@ -17,207 +17,41 @@ limitations under the License.
-			expectedError:    nil,
-		},
+	if err := CreateOrUpdateCertificates(&cfg, "dummyclustername"); err != nil {
+		t.Fatalf("Error creating certificates")
```
```suggestion
		t.Errorf("Error creating certificates")
```

tariq1890 commented on pkg/cloud/azure/services/certificates/certificates_test.go:
```diff
@@ -17,207 +17,41 @@ limitations under the License.
-		},
+	if err := CreateOrUpdateCertificates(&cfg, "dummyclustername"); err != nil {
+		t.Fatalf("Error creating certificates")
+		return
```
```suggestion

```

tariq1890 commented on pkg/cloud/azure/services/certificates/certificates_test.go:
```diff
@@ -17,207 +17,41 @@ limitations under the License.
-				if expectedCommonName != actualCommonName {
-					return errors.Errorf("Unexpected CommonName, Want: [%s]; Got: [%s]", expectedCommonName, actualCommonName)
-				}
+	if !cfg.SAKeyPair.HasCertAndKey() {
```
Same as previous suggestion.

tariq1890 commented on pkg/cloud/azure/services/certificates/certificates_test.go:
```diff
@@ -17,207 +17,41 @@ limitations under the License.
-				if !cert.IsCA {
-					return errors.Errorf("Unexpected value for IsCA, Want: [true]; Got: [%t]", cert.IsCA)
-				}
+	if !cfg.EtcdCAKeyPair.HasCertAndKey() {
```
ditto

tariq1890 commented on pkg/cloud/azure/services/certificates/certificates_test.go:
```diff
@@ -17,207 +17,41 @@ limitations under the License.
-				return nil
-			},
-		},
+	if cfg.AdminKubeconfig == "" {
```
ditto

tariq1890 commented on pkg/cloud/azure/services/certificates/certificates_test.go:
```diff
@@ -17,207 +17,41 @@ limitations under the License.
-		if err := tc.testFunc(); err != nil {
-			t.Fatalf("[%s] failed: %v", tc.name, err)
-		}
+	if len(cfg.DiscoveryHashes) <= 0 {
```
ditto

tariq1890 commented on Gopkg.lock:
```diff
@@ -113,6 +113,14 @@
   version = "v2.9.0"
 
+[[projects]]
+  digest = "1:f1f2bd73c025d24c3b93abf6364bccb802cf2fdedaa44360804c67800e8fab8d"
```
Why do we need this?

awesomenix replied:
Really not sure, i just ran dep ensure and it seems to have updated this.must have been a transitive dependency

tariq1890 commented on pkg/apis/azureprovider/v1alpha1/zz_generated.deepcopy.go:
```diff
@@ -34,6 +34,11 @@ func (in *AzureClusterProviderSpec) DeepCopyInto(out *AzureClusterProviderSpec)
 	in.EtcdCAKeyPair.DeepCopyInto(&out.EtcdCAKeyPair)
 	in.FrontProxyCAKeyPair.DeepCopyInto(&out.FrontProxyCAKeyPair)
 	in.SAKeyPair.DeepCopyInto(&out.SAKeyPair)
+	if in.DiscoveryHashes != nil {
```
This is more of a question? Is this how we deepcopy a string slice in Go?

awesomenix replied:
Interesting code, this is part of auto generation, interesting that it generates that piece of code :)

justaugustus replied:
Yep, the deepcopy is generated, so no need to review that.

tariq1890 commented on Gopkg.lock:
```diff
@@ -837,22 +845,23 @@
-  revision = "05914d821849570fba9eacfb29466f2d8d3cd229"
-  version = "kubernetes-1.13.1"
+  revision = "74b699b93c15473932b89e3d1818ba8282f3b5ab"
+  version = "kubernetes-1.13.3"
```
Kinda ambivalent towards locking into a Kubernetes dependency without pinning it. Will defer to @justaugustus on that.

If this is agreed upon, Can we change to `kubernetes-1.13.4`

justaugustus commented on Gopkg.toml:
```diff
@@ -48,15 +48,6 @@ required = [
-#  - https://github.com/kubernetes-sigs/controller-runtime/pull/349
-[[constraint]]
-  name = "k8s.io/client-go"
-  version = "kubernetes-1.13.1"
```
Any reason for removing this constraint here?

awesomenix replied:
This was in response to Tariq's concern here https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/136#discussion_r265398524

awesomenix replied:
I am actually torn between pinning vs non-pinning, but understand the need to just relying on upstream dependencies rather than putting a constraint or an override here, what do you think?

detiber replied:
Constraining or overriding here could very well lead to conflicts with the controller-runtime and controller-tools dependencies.

justaugustus commented:
/cc @detiber @chuckha @vincepri @randomvariable 
(for opinions)

detiber commented on pkg/apis/azureprovider/v1alpha1/azureclusterproviderconfig_types.go:
```diff
@@ -48,6 +48,9 @@ type AzureClusterProviderSpec struct {
 	// SAKeyPair is the service account key pair.
 	SAKeyPair KeyPair `json:"saKeyPair,omitempty"`
 
+	AdminKubeconfig string `json:"adminKubeconfig,omitempty"`
```
I would consider moving `AdminKubeconfig` to the Status if you feel it is necessary to add it. Otherwise it could be generated on-demand from the `CAKeyPair`.

awesomenix replied:
Thanks it makes sense, updated PR

detiber commented on pkg/apis/azureprovider/v1alpha1/azureclusterproviderconfig_types.go:
```diff
@@ -48,6 +48,9 @@ type AzureClusterProviderSpec struct {
 
+	AdminKubeconfig string `json:"adminKubeconfig,omitempty"`
+
+	DiscoveryHashes []string `json:"discoveryHashes,omitempty"`
```
Similar to my comment about `AdminKubeconfig`, if there is a need to be able to reference this without generating it on demand, consider moving this to Status.

awesomenix replied:
Thanks it makes sense, updated PR

detiber commented:
@vincepri we should definitely sync on these changes with the AWS provider.

awesomenix commented:
@detiber @justaugustus @juan-lee @tariq1890 Thanks for helpful comments, I have updated the PR to incorporate all the comments, would appreciate another look at PR

justaugustus commented:
/lgtm
/approve

justaugustus commented:
/hold cancel
